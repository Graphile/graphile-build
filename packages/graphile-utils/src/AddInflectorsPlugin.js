export default function AddInflectorsPlugin(additionalInflectors) {
  return builder => {
    builder.hook("inflection", (inflection, build) => {
      return build.extend(inflection, additionalInflectors);
    });
  };
}
