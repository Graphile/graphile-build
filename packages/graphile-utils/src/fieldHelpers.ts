import { GraphQLResolveInfo } from "graphql";
import { Build, Context } from "graphile-build";
import { QueryBuilder, SQL } from "graphile-build-pg";

export type SelectGraphQLResultFromTable = (
  tableFragment: SQL,
  builderCallback: (alias: SQL, sqlBuilder: QueryBuilder) => void
) => Promise<any>;

export type GraphileHelpers<TSource> = {
  fieldContext: Context<TSource>;
  selectGraphQLResultFromTable: SelectGraphQLResultFromTable;
};

export function makeFieldHelpers<TSource>(
  build: Build,
  fieldContext: Context<TSource>,
  context: any,
  resolveInfo: GraphQLResolveInfo
) {
  const { parseResolveInfo, pgQueryFromResolveData, pgSql: sql } = build;
  const { getDataFromParsedResolveInfoFragment } = fieldContext;
  const selectGraphQLResultFromTable: SelectGraphQLResultFromTable = async (
    tableFragment: SQL,
    builderCallback: (alias: SQL, sqlBuilder: QueryBuilder) => void
  ) => {
    const { pgClient } = context;
    const parsedResolveInfoFragment = parseResolveInfo(resolveInfo);
    const PayloadType = resolveInfo.returnType;
    const resolveData = getDataFromParsedResolveInfoFragment(
      parsedResolveInfoFragment,
      PayloadType
    );
    const tableAlias = sql.identifier(Symbol());
    const query = pgQueryFromResolveData(
      tableFragment,
      tableAlias,
      resolveData,
      {},
      (sqlBuilder: QueryBuilder) => builderCallback(tableAlias, sqlBuilder)
    );
    const { text, values } = sql.compile(query);
    const { rows } = await pgClient.query(text, values);
    return rows;
  };

  const graphileHelpers: GraphileHelpers<TSource> = {
    fieldContext,
    selectGraphQLResultFromTable,
  };
  return graphileHelpers;
}
