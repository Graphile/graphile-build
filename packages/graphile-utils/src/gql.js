import { parse, visit } from "graphql";
const $$embed = Symbol("graphile-embed");

export function isEmbed(obj) {
  return obj && obj[$$embed] === true;
}

export function embed(value) {
  return {
    [$$embed]: true,
    value: {
      kind: "GraphileEmbed",
      value,
    },
  };
}

export function gql(strings, ...interpolatedValues) {
  const gqlStrings = [];
  const placeholders = {};
  const createPlaceholderFor = value => {
    const rand = String(Math.random());
    placeholders[rand] = value;
    return `"${rand}"`;
  };
  for (let idx = 0, length = strings.length; idx < length; idx++) {
    gqlStrings.push(strings[idx]);
    if (idx === length - 1) {
      // NOOP: last string, so no matching interpolatedValue.
    } else {
      const interpolatedValue = interpolatedValues[idx];
      if (isEmbed(interpolatedValue)) {
        gqlStrings.push(createPlaceholderFor(interpolatedValue.value));
      } else {
        if (typeof interpolatedValue !== "string") {
          throw new Error(
            `Placeholder ${idx +
              1} is invalid - expected string, but received '${typeof interpolatedValue}'. Happened after '${gqlStrings.join(
              ""
            )}'`
          );
        }
        gqlStrings.push(String(interpolatedValue));
      }
    }
  }
  const ast = parse(gqlStrings.join(""));
  const visitor = {
    enter(node, _key, _parent, _path, _ancestors) {
      if (node.kind === "Argument") {
        if (node.value.kind === "StringValue") {
          if (placeholders[node.value.value]) {
            return {
              ...node,
              value: placeholders[node.value.value],
            };
          }
        }
      }
      return undefined;
    },
  };
  const astWithPlaceholdersReplaced = visit(ast, visitor);
  return astWithPlaceholdersReplaced;
}
