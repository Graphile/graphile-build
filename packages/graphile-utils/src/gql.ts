import { parse, visit, ASTNode } from "graphql";
const $$embed = Symbol("graphile-embed");

export interface GraphileEmbed<T = any> {
  [$$embed]: true;
  kind: "GraphileEmbed";
  value: T;
}

export function isEmbed(obj: any): obj is GraphileEmbed {
  return obj && obj[$$embed] === true;
}

export function embed<T>(value: T): GraphileEmbed<T> {
  return {
    [$$embed]: true,
    kind: "GraphileEmbed",
    value,
  };
}

export function gql(
  strings: TemplateStringsArray,
  ...interpolatedValues: Array<string | GraphileEmbed>
) {
  const gqlStrings = [];
  const placeholders = {};
  const createPlaceholderFor = (value: any) => {
    const rand = String(Math.random());
    placeholders[rand] = value;
    return `"${rand}"`;
  };
  for (let idx = 0, length = strings.length; idx < length; idx++) {
    gqlStrings.push(strings[idx]);
    if (idx === length - 1) {
      // NOOP: last string, so no matching interpolatedValue.
    } else {
      const interpolatedValue = interpolatedValues[idx];
      if (isEmbed(interpolatedValue)) {
        gqlStrings.push(createPlaceholderFor(interpolatedValue));
      } else {
        if (typeof interpolatedValue !== "string") {
          throw new Error(
            `Placeholder ${idx +
              1} is invalid - expected string, but received '${typeof interpolatedValue}'. Happened after '${gqlStrings.join(
              ""
            )}'`
          );
        }
        gqlStrings.push(String(interpolatedValue));
      }
    }
  }
  const ast = parse(gqlStrings.join(""));
  const visitor = {
    enter: (node: ASTNode) => {
      if (node.kind === "Argument") {
        if (node.value.kind === "StringValue") {
          if (placeholders[node.value.value]) {
            return {
              ...node,
              value: placeholders[node.value.value],
            };
          }
        }
      }
      return undefined;
    },
  };
  const astWithPlaceholdersReplaced = visit(ast, visitor);
  return astWithPlaceholdersReplaced;
}
