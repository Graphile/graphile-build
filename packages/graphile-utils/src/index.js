import graphqlTag from "graphql-tag";

const $$embed = Symbol("graphile-embed");

export function isEmbed(obj) {
  return obj && obj[$$embed] === true;
}

export function embed(value) {
  return {
    [$$embed]: true,
    value,
  };
}

export function gql(strings, ...placeholders) {
  const gqlStrings = [];
  const gqlValues = [];
  let currentString = "";
  for (let idx = 0, length = strings.length; idx < length; idx++) {
    currentString += strings[idx];
    if (idx === length - 1) {
      gqlStrings.push(currentString);
    } else {
      if (isEmbed(placeholders[idx])) {
        gqlValues.push(placeholders[idx].value);
        gqlStrings.push(currentString);
        currentString = "";
      } else {
        if (typeof placeholders[idx] !== "string") {
          throw new Error(
            `Placeholder ${idx +
              1} is invalid - expected string, but received '${typeof placeholders[
              idx
            ]}'. Happened after '${currentString}'`
          );
        }
        currentString += String(placeholders[idx]);
      }
    }
  }
  return graphqlTag(gqlStrings, ...gqlValues);
}

export function ExtendSchemaPlugin(generator) {
  function getName(name) {
    if (name && name.kind === "Name" && name.value) {
      return name.value;
    }
    throw new Error("Could not extract name from AST");
  }

  function getDescription(desc) {
    if (!desc) {
      return null;
    } else if (desc.kind === "StringValue") {
      return desc.value;
    } else {
      throw new Error(
        `AST issue, we weren't expecting a description of kind '${
          desc.kind
        }' - PRs welcome!`
      );
    }
  }

  function getType(type, build) {
    if (type.kind === "NamedType") {
      return build.getTypeByName(getName(type.name));
    } else if (type.kind === "NonNullType") {
      return new build.graphql.GraphQLNonNull(getType(type.type, build));
    } else if (type.kind === "ListType") {
      return new build.graphql.GraphQLList(getType(type.type, build));
    } else {
      throw new Error(
        `We don't support AST type definition of kind '${
          type.kind
        }' yet... PRs welcome!`
      );
    }
  }

  function getInterfaces(interfaces, build) {
    if (interfaces.length) {
      throw new Error(
        `We don't support interfaces via ExtendSchemaPlugin yet; PRs welcome!`
      );
    }
    return [];
  }

  function getArguments(args, build) {
    if (args.length) {
      throw new Error(
        `We don't support interfaces via ExtendSchemaPlugin yet; PRs welcome!`
      );
    }
    return {};
  }

  return builder => {
    // Add stuff to the schema
    builder.hook("build", build => {
      const { typeDefs, resolvers = {} } = generator(build);
      if (!typeDefs || !typeDefs.kind === "Document") {
        throw new Error(
          "The first argument to ExtendSchemaPlugin must be generated by the `gql` helper"
        );
      }
      const objectExtensions = {};
      typeDefs.definitions.forEach(definition => {
        if (definition.kind === "ObjectTypeExtension") {
          const name = getName(definition.name);
          if (!objectExtensions[name]) {
            objectExtensions[name] = [];
          }
          objectExtensions[name].push(definition);
        } else {
          throw new Error(
            `Unexpected '${
              definition.kind
            }' definition; we were expecting 'ObjectTypeExtension', i.e. something like 'extend type Foo { ... }'`
          );
        }
      });
      return build.extend(build, {
        ExtendSchemaPlugin_objectExtensions: objectExtensions,
        ExtendSchemaPlugin_resolvers: resolvers,
      });
    });
    //const interfaces = getInterfaces(extension.interfaces, build);

    builder.hook("GraphQLObjectType:fields", (fields, build, context) => {
      const {
        extend,
        ExtendSchemaPlugin_objectExtensions: objectExtensions,
        ExtendSchemaPlugin_resolvers: resolvers,
      } = build;
      const { Self, fieldWithHooks } = context;
      if (objectExtensions[Self.name]) {
        const newFields = objectExtensions[Self.name].reduce(
          (memo, extension) => {
            const fields = extension.fields;
            if (fields && fields.length) {
              fields.forEach(field => {
                if (field.kind === "FieldDefinition") {
                  const description = getDescription(field.description);
                  const fieldName = getName(field.name);
                  const args = getArguments(field.arguments, build);
                  const type = getType(field.type, build);
                  const scope = {
                    // TODO: add directives here!
                  };
                  const deprecationReason = null; // TODO: extract from directive!
                  const resolve =
                    (resolvers[Self.name] && resolvers[Self.name][fieldName]) ||
                    null;
                  memo[fieldName] = fieldWithHooks(
                    fieldName,
                    {
                      type,
                      args,
                      resolve,
                      deprecationReason,
                      description,
                    },
                    scope
                  );
                } else {
                  throw new Error(
                    `AST issue: expected 'FieldDefinition', instead received '${
                      field.kind
                    }'`
                  );
                }
              });
            }
            return memo;
          },
          {}
        );
        // EXTEND!
        return extend(fields, newFields);
      } else {
        return fields;
      }
    });
  };
}
