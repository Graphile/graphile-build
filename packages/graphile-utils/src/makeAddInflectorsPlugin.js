export default function makeAddInflectorsPlugin(additionalInflectors) {
  return builder => {
    builder.hook("inflection", (inflection, build) => {
      return build.extend(inflection, additionalInflectors);
    });
  };
}
