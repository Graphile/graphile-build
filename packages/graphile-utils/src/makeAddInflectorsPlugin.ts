import { Plugin } from "graphile-build";

export default function makeAddInflectorsPlugin(
  additionalInflectors: {
    [str: string]: ((...args: Array<any>) => any);
  },
  replace = false
): Plugin {
  return builder => {
    builder.hook("inflection", (inflection, build) => {
      if (replace) {
        // Overwrite directly so that we don't lose the 'extend' hints
        Object.assign(inflection, additionalInflectors);
        return inflection;
      } else {
        return build.extend(
          inflection,
          additionalInflectors,
          `Adding inflectors ('${Object.keys(additionalInflectors).join(
            "', '"
          )}') via makeAddInflectorsPlugin. You can pass \`true\` as the second argument to makeAddInflectorsPlugin to allow overwriting existing inflectors.`
        );
      }
    });
  };
}
