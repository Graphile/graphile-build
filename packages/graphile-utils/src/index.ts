import { embed, gql } from "./gql";
import makeAddInflectorsPlugin from "./makeAddInflectorsPlugin";
import makeExtendSchemaPlugin from "./makeExtendSchemaPlugin";
import makePluginByCombiningPlugins from "./makePluginByCombiningPlugins";
import makeWrapResolversPlugin from "./makeWrapResolversPlugin";

export {
  embed,
  gql,
  makeAddInflectorsPlugin,
  makeExtendSchemaPlugin,
  makePluginByCombiningPlugins,
  makeWrapResolversPlugin,
};
