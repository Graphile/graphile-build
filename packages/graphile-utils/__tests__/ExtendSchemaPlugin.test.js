import { ExtendSchemaPlugin, gql } from "../src";
import {
  buildSchema,
  defaultPlugins,
  StandardTypesPlugin,
  QueryPlugin,
  MutationPlugin,
  MutationPayloadQueryPlugin,
} from "graphile-build";
import { graphql, printSchema } from "graphql";

const simplePlugins = [
  StandardTypesPlugin,
  QueryPlugin,
  MutationPlugin,
  MutationPayloadQueryPlugin,
];

it("allows adding a simple type", async () => {
  const schema = await buildSchema([
    ...simplePlugins,
    ExtendSchemaPlugin(build => ({
      typeDefs: gql`
        extend type Query {
          """
          comment
          """
          randomNumber: Int
        }
      `,
      resolvers: {
        Query: {
          randomNumber(_query, _args, _context, _info) {
            return 4; // chosen by fair dice roll. guaranteed to be random. xkcd#221
          },
        },
      },
    })),
  ]);
  const printedSchema = printSchema(schema);
  expect(printedSchema).toMatchSnapshot();
  const { data } = await graphql(
    schema,
    `
      {
        randomNumber
      }
    `
  );
  expect(data.randomNumber).toEqual(4);
});
