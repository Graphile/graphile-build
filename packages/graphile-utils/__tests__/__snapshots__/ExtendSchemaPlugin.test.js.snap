// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows adding a field with arguments 1`] = `
"\\"\\"\\"The root query type which gives access points into the data universe.\\"\\"\\"
type Query {
  \\"\\"\\"
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  \\"\\"\\"
  query: Query!

  \\"\\"\\"Gives you back what you put in\\"\\"\\"
  echo(input: [Int!]!): [Int!]!
}
"
`;

exports[`allows adding a field with arguments named using a custom inflector 1`] = `
"\\"\\"\\"The root query type which gives access points into the data universe.\\"\\"\\"
type Query {
  \\"\\"\\"
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  \\"\\"\\"
  query: Query!

  \\"\\"\\"Gives you back what you put in\\"\\"\\"
  myCustomEchoFieldName(input: [Int!]!): [Int!]!
}
"
`;

exports[`allows adding a non-null list of non-null type 1`] = `
"\\"\\"\\"The root query type which gives access points into the data universe.\\"\\"\\"
type Query {
  \\"\\"\\"
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  \\"\\"\\"
  query: Query!

  \\"\\"\\"
  Gives a list of numbers that were randomly generated by fair dice roll
  \\"\\"\\"
  randomNumbers: [Int!]!
}
"
`;

exports[`allows adding a non-null type 1`] = `
"\\"\\"\\"The root query type which gives access points into the data universe.\\"\\"\\"
type Query {
  \\"\\"\\"
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  \\"\\"\\"
  query: Query!

  \\"\\"\\"A random number generated by a fair dice roll.\\"\\"\\"
  randomNumber: Int!
}
"
`;

exports[`allows adding a simple type 1`] = `
"\\"\\"\\"The root query type which gives access points into the data universe.\\"\\"\\"
type Query {
  \\"\\"\\"
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  \\"\\"\\"
  query: Query!

  \\"\\"\\"A random number generated by a fair dice roll.\\"\\"\\"
  randomNumber: Int
}
"
`;
